i2c: i2c_mcux_flexcomm: add multi-target support

bravl opened the pull request:
Add support for multiple target addresses. The flexcomm peripheral support up-to 4 target addresses per interface. This commit extends the target implementation to allow the registration of multiple targets.

This is done by splitting the target configuration from the peripheral configuration. When the data callback is triggered, the destination address is used to route retrieve the correct target configuration and call the associated callbacks.

Tested on LPC55S69

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -193,23 +203,50 @@ static int mcux_flexcomm_transfer(const struct device *dev,
+	return -EBUSY;
+}
+
+static int mcux_flexcomm_target_find_target_by_addr(struct mcux_flexcomm_data *data,
```
find_target_... sounds like this should maybe return a target (or a NULL pointer) instead of the index?

jacmet replied:
Isn't that backwards? The index is not really interesting, how about making _find_free_target() and and _find_target_by_addr simply return a pointer to struct mcux_flexcomm_target_config or NULL?

bravl replied:
reworked to `_find_free_target`

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -193,23 +203,50 @@ static int mcux_flexcomm_transfer(const struct device *dev,
+	int target_idx = mcux_flexcomm_target_find_target_by_addr(userData, address);
 	struct mcux_flexcomm_data *data = userData;
-	const struct i2c_target_callbacks *target_cb = data->target_cfg->callbacks;
+	struct mcux_flexcomm_target_config *target = &data->targets[target_idx];
```
target_idx may be -ENODEV here

bravl replied:
Reworked

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -193,23 +203,50 @@ static int mcux_flexcomm_transfer(const struct device *dev,
+					     uint16_t addr)
+{
+	for (int i = 0; i < MCUX_FLEXCOMM_MAX_SLAVES; i++) {
+		if (data->targets[i].target_cfg->address == addr) {
```
this should probably also check target_attached to ensure we don't get a stale entry (E.G. you don't clear address when freeing).

I think the change would be easier to read if you split it up in a small series, E.G.:
- add struct mcux_flexcomm_target_config and update code to use target pointer
- add (dummy) _find_free_target / _find_target_by_addr
- support up to 4 slave addresses

bravl replied:
Reworked to split into seperate commits as suggested.
Also added proper unregistering and cleanup.

bravl committed:
i2c: i2c_mcux_flexcomm: move target data to separate struct

Move target data to dedicated struct. This is in preparation for multi
target support. The target_handle can stay global since this handle is
unique per peripheral and not per slave address.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

bravl committed:
i2c: i2c_mcux_flexcomm: add find target functions

Add helper functions to find a free target for configuration and find a
target based on the configured slave address. These functions are in
preparation for multi-target mode.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

bravl commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -200,16 +200,48 @@ static int mcux_flexcomm_transfer(const struct device *dev,
 
 #if defined(CONFIG_I2C_TARGET)
+
+struct mcux_flexcomm_target_data *_find_free_target_data(struct mcux_flexcomm_data *data)
```
Fixed function naming, added `mcux_flexcomm_` prefix.

bravl commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -203,25 +206,73 @@ static int mcux_flexcomm_transfer(const struct device *dev,
 	return NULL;
 }
 
+int _target_to_idx(struct mcux_flexcomm_data *data, struct mcux_flexcomm_target_data *target)
```
Correct, that's the only used for the function. Moved the logic inside the `mcux_flexcomm_setup_i2c_config_address`.

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -200,16 +200,48 @@ static int mcux_flexcomm_transfer(const struct device *dev,
 
 #if defined(CONFIG_I2C_TARGET)
+
+struct mcux_flexcomm_target_data *mcux_flexcomm_find_free_target(struct mcux_flexcomm_data *data)
```
how come these are not static?

bravl replied:
Correct, fixed.

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -204,25 +207,76 @@ static int mcux_flexcomm_transfer(const struct device *dev,
 {
-	struct mcux_flexcomm_target_data *target = &data->target_data;
+	int i;
+	struct mcux_flexcomm_target_data *target;
```
NIT: the driver uses reverse christmas tree order, so please put int i after target (here and below)

bravl replied:
FIxed

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -204,25 +207,76 @@ static int mcux_flexcomm_transfer(const struct device *dev,
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(data->target_data); i++) {
+		if (&data->target_data[i] == target) {
```
NIT: you are not checking target_attached here

bravl replied:
Added check.

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -204,25 +207,76 @@ static int mcux_flexcomm_transfer(const struct device *dev,
 }
 
+static int mcux_flexcomm_setup_i2c_config_address(struct mcux_flexcomm_data *data,
+		struct mcux_flexcomm_target_data *target, uint8_t address, bool disabled)
```
NIT: target already contains the address, so why pass it separately?

bravl replied:
Indeed, removed address parameter.

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -345,18 +404,39 @@ int mcux_flexcomm_target_unregister(const struct device *dev,
+				i2c_target_transfer_callback, data);
+		I2C_SlaveTransferNonBlocking(base, &data->target_handle,
+				kI2C_SlaveCompletionEvent | kI2C_SlaveTransmitEvent |
+				kI2C_SlaveReceiveEvent | kI2C_SlaveDeselectedEvent);
```
NIT: this logic could be moved to a helper function so it can be called both from here and from mcux_flexcomm_target_register() to ensure they stay in sync

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -345,18 +423,36 @@ int mcux_flexcomm_target_unregister(const struct device *dev,
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys,
+				   &clock_freq)) {
+		return -EINVAL;
+	}
```
This looks unused?

bravl replied:
removed dead code.

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -292,6 +346,29 @@ static void i2c_target_transfer_callback(I2C_Type *base,
 	}
 }
 
+static int mcux_flexcomm_setup_slave_config(const struct device *dev)
```
NIT: The diff would be a bit easier to read if you added this helper in a separate commit before adding the multi-slave support

bravl replied:
Moved to separate commit

jacmet commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -345,18 +423,36 @@ int mcux_flexcomm_target_unregister(const struct device *dev,
+		if (mcux_flexcomm_setup_slave_config(dev) < 0) {
+			return -EINVAL;
+		}
+	}
```
NIT: you have the special casing in the else clause, it IMHO would be more logical to swap it around, E.G.
```
if (data->nr_targets_attached > 0) {
  	/* still slaves attached, reconfigure the I2C peripheral after address removal */
	if (mcux_flexcomm_setup_slave_config(dev) < 0) {
		return -EINVAL;
	}
} else {
       I2C_SlaveDeinit(base);
}
```

bravl committed:
i2c: i2c_mcux_flexcomm: move slave init to helper function

Move slave initialization to a separate helper function. This helper will
also be needed in the multi target support during unregistration. When
unregistering, the slave needs to be reinitialized if other remaining
targets are still attached.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

bravl committed:
i2c: i2c_mcux_flexcomm: add multi target support

When registering a target device, the driver will check if a slave address
is free in the peripheral. If so, the new slave address gets configured and
enabled.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

jacmet reviewed (approved):
LGTM, thanks!

teburd commented:
One note that it would be nice if the title of the PR matched the commit terminology, the commit notes its a multi-target support while the PR title terms use slave which the I2C spec has moved away from

bravl commented:
> One note that it would be nice if the title of the PR matched the commit terminology, the commit notes its a multi-target support while the PR title terms use slave which the I2C spec has moved away from

Updated title and description to match commit terminology.

teburd commented on drivers/i2c/i2c_mcux_flexcomm.c:
```diff
@@ -23,6 +23,8 @@ LOG_MODULE_REGISTER(mcux_flexcomm);
 	COND_CODE_0(CONFIG_I2C_NXP_TRANSFER_TIMEOUT, (K_FOREVER),                                  \
 		    (K_MSEC(CONFIG_I2C_NXP_TRANSFER_TIMEOUT)))
 
+#define MCUX_FLEXCOMM_MAX_TARGETS 4
```
I guess ideally this would be part of device tree

bravl replied:
> I guess ideally this would be part of device tree

I could add something like `num-targets` or `max-targets` to the lpc-i2c?

jacmet replied:
Does the number of target devices differ between the SoCs using this driver? If not, why put it in the device tree?

bravl replied:
I checked the LPC55xxx, LPC54xxx datasheets and the hal_nxp and they all seem to refer to 4 target addresses.

danieldegrasse replied:
@DerekSnell are you aware of any instances of the Flexcomm IP that support more than 4 target addresses? The RT595 also seems to only support 4

teburd replied:
Ah then yeah this seems ok. The naming of this, without looking at the IP suggested it was a maximum and there may be smaller numbers for some IP instances perhaps. Ignore this DT comment in that case.
